### PR TITLE
Wrap xla_quantized_matmul with shard map

### PR DESCRIPTION
**Pull request by rupengliu-meta**
# Description
This is to remove all reduce triggered by reduce_max. This also implicitly converts the quantization from per channel into block
截屏2026-01-26 下午12 36 01
This is the latest trace and I can see that the all-reduce is gone
截屏2026-01-26 下午7 21 42

Perf improved around: 7%
截屏2026-01-27 上午10 12 29
截屏2026-01-27 上午10 12 38

The rest of the description includes relevant details and context, examples:
- why is this change being made,
- the problem being solved and any relevant context,
- why this is a good solution,
- some information about the specific implementation,
- shortcomings of the solution and possible future improvements.

If the change fixes a Github issue, please include a link, e.g.,:
FIXES: #123456

# Tests

Please describe how you tested this change, and include any instructions and/or
commands to reproduce.

# Checklist

Before submitting this PR, please make sure:
- I have performed a self-review of my code.
- I have necessary comments in my code, particularly in hard-to-understand areas.
- I have made or will make corresponding changes to any relevant documentation.

### CODE REVIEW

File: tpu_inference/layers/vllm/linear.py [hunk 1]
```diff
@@ -40,27 +40,27 @@ def sharded_quantized_matmul(x: jax.Array, w_q: jax.Array, w_s: jax.Array,
 
     # NOTE (jacobplatin/kyuyeunk) there have been numeric issues (concerning) NaNs
     # with the kernel and thus we disable it for now.
-    if envs.ENABLE_QUANTIZED_MATMUL_KERNEL:
```
Comment (vanbasten23) [inline tpu_inference/layers/vllm/linear.py:43]: Good catch

Comment (kyuyeunk) [reply tpu_inference/layers/vllm/linear.py:43]: yes, and this code change maintains the behavior (see line 52). only difference is that xla based kernel is executed inside `shard_map` to get rid of `all reduce`. 

previously, it was doing per-channel quantization. and the issue was that if `x=(b,d)` and `x` was sharded along `d` dimension, it will trigger allreduce during `jnp.max(x, axis=1)` but if this is executed inside `shard_map`, no collective operation is needed. also, we get extra accuracy boost for free because we are essentially doing block, quantization where `num_blocks=tp_size`.

**Comment by kyuyeunk** [comment]
in the description, can you add e2e performance difference between before and after this change?

**Comment by vanbasten23** [comment]
By "all reduce after the reduce_max", do you mean the `all-reduce.370`?

File: tpu_inference/layers/vllm/linear.py (hunk 1 above)
Comment (vanbasten23) [reply tpu_inference/layers/vllm/linear.py:43]: Thanks @kyuyeunk for the explanation! Makes sense.

**Comment by kyuyeunk** [comment]
> By "all reduce after the reduce_max", do you mean the `all-reduce.370`?

@rupengliu-meta probably worth renaming the description into "all reduce triggered by reduce_max".

**Comment by rupengliu-meta** [comment]
> > By "all reduce after the reduce_max", do you mean the `all-reduce.370`?
> 
> @rupengliu-meta probably worth renaming the description into "all reduce triggered by reduce_max".

updated

**Comment by rupengliu-meta** [comment]
> By "all reduce after the reduce_max", do you mean the `all-reduce.370`?

yes

**Comment by rupengliu-meta** [comment]
> in the description, can you add e2e performance difference between before and after this change?

will update tmr!

**Comment by rupengliu-meta** [comment]
updated with perf gain, thanks!
